akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1245,7 +1246,7 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   private var callbacksWaitingForInterpreter: List[ConcurrentAsyncCallback[_]] = Nil
   // is used for two purposes: keep track of running callbacks and signal that the
   // stage has stopped to fail incoming async callback invocations by being set to null
-  private val asyncCallbacksInProgress = new AtomicReference[List[Promise[Done]]](Nil)
+  private val asyncCallbacksInProgress = new AtomicReference(new ConcurrentHashMap[Promise[Done], Promise[Done]]())
```

Comment:
A straight-forward alternative (but broken) solution would be to use a `val stageCompletedPromise = Promise[Done]` and then use `promise.tryCompleteWith(stageCompletedPromise)` above to race completion as done in `GraphInterpreter.runAsyncInput` against completion of the stage.

Here's the catch: `promise.tryCompleteWith(stageCompletedPromise)` would create a reference from the `stageCompletedPromise` to the callback `promise`. In the normal case, the callback `promise` would however be completed by `GraphInterpreter.runAsyncInput`. In that case, any references should be dropped afterwards, however, `stageCompletedPromise` would keep references to that promise leading to another kind of memory leak. That's why we seem to need another kind of tracking mechanism that avoids keeping callback promises referenced longer-term.